File: src/learning/translation_memory.js
/**
 * Translation Memory
 * Manages stored translations and provides retrieval for similar texts
 */

class TranslationMemory {
  constructor() {
    this.memory = new Map(); // Stores source -> target translation pairs
    this.totalTranslations = 0;
    this.accessCount = new Map(); // Tracks how often each translation is used
  }

  /**
   * Store a translation pair
   * @param {string} source - Source text
   * @param {string} target - Target translation
   */
  store(source, target) {
    // Use the source text as a key to store the translation
    this.memory.set(source, target);
    this.totalTranslations++;
    
    // Initialize access count for this translation
    if (!this.accessCount.has(source)) {
      this.accessCount.set(source, 0);
    }
    
    console.log(`Stored translation pair (${this.totalTranslations} total)`);
  }

  /**
   * Retrieve a translation for source text
   * @param {string} source - Source text to translate
   * @returns {string|null} - Translation if found, null otherwise
   */
  retrieve(source) {
    const translation = this.memory.get(source);
    
    if (translation) {
      // Update access count
      const currentCount = this.accessCount.get(source) || 0;
      this.accessCount.set(source, currentCount + 1);
    }
    
    return translation;
  }

  /**
   * Find similar translations using string similarity
   * @param {string} source - Source text to find similar translations for
   * @param {number} threshold - Similarity threshold (0-1)
   * @returns {Array} - Array of similar translation pairs
   */
  findSimilar(source, threshold = 0.8) {
    const results = [];
    const sourceLower = source.toLowerCase();
    
    for (const [storedSource, target] of this.memory) {
      const similarity = this.calculateSimilarity(sourceLower, storedSource.toLowerCase());
      
      if (similarity >= threshold) {
        results.push({
          source: storedSource,
          target: target,
          similarity: similarity
        });
      }
    }
    
    // Sort by similarity (highest first)
    results.sort((a, b) => b.similarity - a.similarity);
    
    return results;
  }

  /**
   * Calculate string similarity using a simple algorithm
   * @param {string} str1 - First string
   * @param {string} str2 - Second string
   * @returns {number} - Similarity score between 0 and 1
   */
  calculateSimilarity(str1, str2) {
    // Simple similarity calculation (could be enhanced with more sophisticated algorithms)
    const longer = str1.length > str2.length ? str1 : str2;
    const shorter = str1.length > str2.length ? str2 : str1;
    
    if (longer.length === 0) {
      return 1.0;
    }
    
    const editDistance = this.levenshteinDistance(longer, shorter);
    return (longer.length - editDistance) / longer.length;
  }

  /**
   * Calculate Levenshtein distance between two strings
   * @param {string} str1 - First string
   * @param {string} str2 - Second string
   * @returns {number} - Edit distance
   */
  levenshteinDistance(str1, str2) {
    const matrix = Array(str2.length + 1).fill().map(() => Array(str1.length + 1).fill(0));
    
    for (let i = 0; i <= str1.length; i++) {
      matrix[0][i] = i;
    }
    
    for (let j = 0; j <= str2.length; j++) {
      matrix[j][0] = j;
    }
    
    for (let j = 1; j <= str2.length; j++) {
      for (let i = 1; i <= str1.length; i++) {
        const cost = str1[i - 1] === str2[j - 1] ? 0 : 1;
        matrix[j][i] = Math.min(
          matrix[j][i - 1] + 1,     // deletion
          matrix[j - 1][i] + 1,     // insertion
          matrix[j - 1][i - 1] + cost // substitution
        );
      }
    }
    
    return matrix[str2.length][str1.length];
  }

  /**
   * Get the size of the translation memory
   * @returns {number} - Number of stored translation pairs
   */
  getSize() {
    return this.memory.size;
  }

  /**
   * Get the total number of translations stored
   * @returns {number} - Total translation count
   */
  getTotalTranslations() {
    return this.totalTranslations;
  }

  /**
   * Clear the translation memory
   */
  clear() {
    this.memory.clear();
    this.accessCount.clear();
    this.totalTranslations = 0;
    console.log('Translation memory cleared');
  }
}

module.exports = TranslationMemory;

File: src/learning/translation_memory.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""In-memory translation memory compatibility layer for legacy JS removal."""

from __future__ import annotations

from dataclasses import dataclass
from difflib import SequenceMatcher


@dataclass(slots=True)
class SimilarTranslation:
    source: str
    target: str
    similarity: float

    def to_dict(self) -> dict:
        return {"source": self.source, "target": self.target, "similarity": self.similarity}


class TranslationMemory:
    """Small in-memory TM; production persistence lives in src.state.translation_memory."""

    def __init__(self):
        self.memory: dict[str, str] = {}
        self.total_translations = 0
        self.access_count: dict[str, int] = {}

    def store(self, source: str, target: str):
        self.memory[source] = target
        self.total_translations += 1
        self.access_count.setdefault(source, 0)

    def retrieve(self, source: str) -> str | None:
        translation = self.memory.get(source)
        if translation is not None:
            self.access_count[source] = self.access_count.get(source, 0) + 1
        return translation

    def find_similar(self, source: str, threshold: float = 0.8) -> list[dict]:
        results = []
        source_key = source.casefold()
        for stored_source, target in self.memory.items():
            similarity = self.calculate_similarity(source_key, stored_source.casefold())
            if similarity >= threshold:
                results.append(SimilarTranslation(stored_source, target, similarity).to_dict())
        return sorted(results, key=lambda item: item["similarity"], reverse=True)

    def calculate_similarity(self, left: str, right: str) -> float:
        return SequenceMatcher(None, left, right).ratio()

    def levenshtein_distance(self, left: str, right: str) -> int:
        if left == right:
            return 0
        if not left:
            return len(right)
        if not right:
            return len(left)
        previous = list(range(len(right) + 1))
        for i, left_char in enumerate(left, start=1):
            current = [i]
            for j, right_char in enumerate(right, start=1):
                cost = 0 if left_char == right_char else 1
                current.append(min(current[j - 1] + 1, previous[j] + 1, previous[j - 1] + cost))
            previous = current
        return previous[-1]

    def get_size(self) -> int:
        return len(self.memory)

    def get_total_translations(self) -> int:
        return self.total_translations

    def clear(self):
        self.memory.clear()
        self.access_count.clear()
        self.total_translations = 0
